src/libexpr/include/nix/expr/value.hh

```diff
@@ -393,12 +436,44 @@ class ValueStorage<ptrSize, std::enable_if_t<detail::useBitPackedValueStorage<pt
     };
 
     using PackedPointer = typename PackedPointerTypeStruct<ptrSize>::type;
-    using Payload = std::array<PackedPointer, 2>;
-    Payload payload = {};
 
-    static constexpr int discriminatorBits = 3;
+    /**
+     * For multithreaded evaluation, we have to make sure that thunks/apps
+     * (the only mutable types of values) are updated in a safe way. A
+     * value can have the following states (see `force()`):
+     *
+     * * "thunk"/"app". When forced, this value transitions to
+     *   "pending". The current thread will evaluate the
+     *   thunk/app. When done, it will override the value with the
+     *   result. If the value is at that point in the "awaited" state,
+     *   the thread will wake up any waiting threads.
+     *
+     * * "pending". This means it's currently being evaluated. If
+     *   another thread forces this value, it transitions to "awaited"
+     *   and the thread will wait for the value to be updated (see
+     *   `waitOnThunk()`).
+     *
+     * * "awaited". Like pending, only it means that there already are
+     *   one or more threads waiting for this thunk.
+     *
+     * To ensure race-free access, the non-atomic word `p1` must
+     * always be updated before `p0`. Writes to `p0` should use
+     * *release* semantics (so that `p1` and any referenced values become
+     * visible to threads that read `p0`), and reads from `p0` should
+     * use `*acquire* semantics.
+     *
+     * Note: at some point, we may want to switch to 128-bit atomics
+     * so that `p0` and `p1` can be updated together
+     * atomically. However, 128-bit atomics are a bit problematic at
+     * present on x86_64 (see
+     * e.g. https://ibraheem.ca/posts/128-bit-atomics/).
+     */
+    std::atomic<PackedPointer> p0{0};
+    PackedPointer p1{0};
+
     static constexpr PackedPointer discriminatorMask = (PackedPointer(1) << discriminatorBits) - 1;
 
+    // FIXME: move/update
```

Comment:
Seems like this could move above the new `enum PrimaryDiscriminator` at the top of this file now